File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.cc
#include <iostream>
#include <vector>
#include <cmath>
#include <algorithm>
using namespace std;

int n, minDiff{ 40000 };
vector<vector<int>> v;

int calc(vector<int>& vec)
{
    int result{ 0 };
    for (int i{ 0 }; i < vec.size(); i++)
    {
        for (int j = i + 1; j < vec.size(); j++)
        {
            result += (v[vec[i]][vec[j]] + v[vec[j]][vec[i]]);
        }
    }
    return result;
}

int main()
{
    cin >> n;
    v.resize(n, vector<int>(n));

    for (int i{ 0 }; i < n; i++)
    {
        for (int j = 0; j < n; j++)
        {
            cin >> v[i][j];
        }
    }

    vector<int> players(n);
    for (int i{ 0 }; i < n; i++)
    {
        players[i] = i;
    }

    vector<bool> comb(n, false);
    for (int i{ n / 2 }; i < n; i++)
    {
        comb[i] = true;
    }

    do
    {
        vector<int> start, link;
        for (int i{ 0 }; i < n; i++)
        {
            if (comb[i])
            {
                start.push_back(i);
            }
            else
            {
                link.push_back(i);
            }
        }
        int diff = abs(calc(start) - calc(link));
        minDiff = min(minDiff, diff);
    } while (next_permutation(comb.begin(), comb.end()));

    cout << minDiff;
}

File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.py
from itertools import combinations


def ability(grid, team):
    total = 0
    for i in range(len(team)):
        for j in range(i + 1, len(team)):  # 중복 연산 방지 (i, j)만 더함
            total += grid[team[i]][team[j]] + grid[team[j]][team[i]]
    return total


n = int(input())
grid = [list(map(int, input().split())) for _ in range(n)]
players = list(range(n))

min_diff = float('inf')  # 최소값을 무한대로 설정

for team in combinations(players, n // 2):  # 조합 사용 → 중복 계산 방지
    other_team = list(set(players) - set(team))  # 여집합 구하기
    diff = abs(ability(grid, team) - ability(grid, other_team))

    min_diff = min(min_diff, diff)

print(min_diff)
